lib/types/record.nix

```diff
@@ -0,0 +1,124 @@
+      wildcard ? null,
+    }@args:
+    # TODO: assert that at least one arg is provided, i.e. args != { }
+    let
+      checkFields = mapAttrs (
+        name: value:
+        if value._type or null != "option" then
+          throw "Record field `${lib.escapeNixIdentifier name}` must be declared with `mkOption`."
+        else if value ? apply then
+          throw "In field ${lib.escapeNixIdentifier name} records do not support options with `apply`"
+        else if value ? readOnly then
+          throw "In field ${lib.escapeNixIdentifier name} records do not support options with `readOnly`"
+        else
+          value
+      );
+      fields' = checkFields fields;
+      # TODO: should we also check optionalFields have no default?
+      optionalFields' = checkFields optionalFields;
+      wildcard =
+        if args.wildcard or null == null then
+          null
+        else if args.wildcard._type or null != "option" then
+          throw "Record wildcard must be declared with `mkOption`."
+        else if args.wildcard ? apply then
+          throw "Records do not support options with `apply`, but wildcard has `apply`."
+        else
+          args.wildcard;
+    in
+    mkOptionType {
+      name = "record";
+      description = if wildcard == null then "record" else "open record of ${wildcard.description}";
+      descriptionClass = if wildcard == null then "noun" else "composite";
+      check = isAttrs;
+      merge =
+        loc: defs:
+        let
+          data = zipAttrsWith (name: values: values) (
+            map (
+              def:
+              mapAttrs (_fieldName: fieldValue: {
+                inherit (def) file;
+                value = fieldValue;
+              }) def.value
+            ) defs
+          );
+          requiredFieldValues = mapAttrs (
+            fieldName: fieldOption:
+            builtins.addErrorContext "while evaluating the field `${fieldName}' of option `${showOption loc}'" (
+              (mergeDefinitions (loc ++ [ fieldName ]) fieldOption.type (
+                data.${fieldName} or [ ]
+                # FIXME: isn't this handled by `mergeDefinitions` already?
+                ++ optional (fieldOption ? default) {
+                  value = fieldOption.default;
+                  file = "the default value of option ${showOption loc}";
+                }
+              )).mergedValue
+            )
+          ) fields';
+          optionalFieldValues = lib.concatMapAttrs (
+            fieldName: fieldOption:
+            builtins.addErrorContext
+              "while evaluating the optional field `${fieldName}' of option `${showOption loc}'"
+              (
+                let
+                  opt = mergeDefinitions (loc ++ [ fieldName ]) fieldOption.type (
+                    data.${fieldName} or [ ]
+                    # TODO: surely defaults make no sense for optional fields?
+                    # And doesn't mergeDefinitions do this internally, anyway?
```

Comment:
It sounds like we need a lazy and non-lazy mode for record types, since its users may wish for either behaviour.

In nixvim, for instance, we probably want the non-lazy behaviour where optional fields are _not_ present in the resulting attrs.

This is because we usually want to be able to recursively map the attrs into a lua table using our version of `lib.generators.toLua`.

I guess we could have our `toLua` generator wrap recursion in `tryEval` and not render anything that throws, but that seems like an overly large hammer for this issue.